single-registry-architecture/README.md

```diff
@@ -0,0 +1,217 @@
+Registry [ACR]) for container images. We use different backends for container images and for helm charts. As a result,
+we have to run synchronization tools to move stuff and replicate across our different registries. For more info and
+diagrams showing how the current solution works, please see
+[here](https://handbook.giantswarm.io/docs/product/architecture-specs-adrs/specs/registry-mirror/#how-are-images-synchronised-between-registries).
+
+While switching away from Docker Hub, we would like also to simplify the architecture, which now involves multiple registries and caring for replication across them. The main reason we started to use replicated registries was
+a requirement of one of our former customers. While this has some benefits in terms of reliability, it also brings
+complexity and cost of running
+the synchronization software on our side. We want to offload the problem of High Availability to a registry provider and
+not to handle it ourselves. To protect us from potential service provider down time or networking issues, we will run a
+local pull-through proxy. This solution should also limit the cost of the remote registry a lot.
+
+One important implication of the proposed change is how it affects our customers. Since now we use the paid subscription
+for Docker Hub, we log in all the container runtimes on all the nodes using our docker hub credentials. This makes
+customers use our subscription to download any images from docker hub: ones that are needed to run our infrastructure,
+as well as any other. After switching to a new architecture, we will care for availability of our images, but leave
+customer's images in the default configuration. This means, that they can start hitting rate limiting where previously
+nothing like that happened. It also means that they might want to configure their own registries in the container
+runtimes to work around this issue.
+
+Using this approach, we will be able to simplify our architecture, limit the cost (Docker Hub is expensive and gets worse)
+and keep HA properties of our infrastructure.
+
+### Specific problems and goals to solve
+
+More specifically, our aim is to:
+
+- Stop using Docker Hub by the end of 2023.
+- Simplify registries architecture.
+- Switch entirely to OCI based registries for both container images and helm charts.
+  - Flux developers have already declared that no new feature will come to Flux for users of HTTP based repos, only for
+  - OCI. Currently, we use a mix of HTTP and OCI repos, but charts discovery uses HTTP based ones only.
+- If possible, switch to single OCI provider.
+- Increase availability and lower the network transfer cost by providing a pull-through cache service.
+- Automatically clean old dev release artifacts.
+- If possible, use the registry provider to replicate registries between Europe and China.
+- Use a solution, that works natively with `containerd` and `helm` and doesn't need tricks like mutating webhooks,
```

Comment:
@silenteh has worked in pull-through cache lately and he can provide good insights about this

Reply:
I think we have 2 case scenarios here:

- moving to a single registry architecture, which is the single source of truth for all GiantSwarm container images, and which is the part you have started the conversation about

- having a pull-through cache for:
    * GiantSwarm container images
    * Customer private registries

Currently I am focussing on the customer private registries, which will require a mutating admission controller.

The mutating admission controller should have the possibility to define the hostname of the container registry that needs to be replaced (actually multiple of them, so a `map[string]string`), so that it acts only on the customer private registries, to start with.

A potential second phase would be to add support for multiple upstream registries in the pull-through cache, so that we can benefit from it, also for the GiantSwarm images.

The idea would be to deploy this app on-demand on the workload clusters (again to start with)

I agree that of all the pull-through caches listed, the docker disribution one is the best, with the caveat that there will be the need to deploy multiple instances, one for each upstream registry

Reply:
In general, I agree, but I think we need the GS images caching solution from day 1, otherwise we risk high cost. Also, I really want to avoid webhooks based approach, as it's a potential failure point, and an ugly one :|